### PR TITLE
Edit user role

### PR DESCRIPTION
**Pull request by montana-mil**
## Description
Allows users with `ccpo`, `owner`, or `admin` permissions to update the role of members in their workspaces. 

Worked with @richard-dds on this.

## Pivotal Tracker
https://www.pivotaltracker.com/story/show/159010528

### CODE REVIEW

File: atst/models/workspace.py [hunk 1]
```diff
@@ -71,4 +72,4 @@ def task_order(self):
 
     @property
     def members(self):
-        return MOCK_MEMBERS
+        return [WorkspaceUser(role.user, role) for role in self.roles]
```
Comment (patricksmithdds) [inline atst/models/workspace.py]: Can we also get rid of `MOCK_MEMBERS` now that it's no longer used anywhere?

File: atst/routes/workspaces.py [hunk 2]
```diff
@@ -119,3 +122,44 @@ def create_member(workspace_id):
...
+def view_member(workspace_id, member_id):
+    workspace = Workspaces.get(g.current_user, workspace_id)
+    Authorization.check_workspace_permission(
+        g.current_user,
+        workspace,
+        Permissions.ASSIGN_AND_UNASSIGN_ATAT_ROLE,
+        "edit this workspace user",
+    )
+    member = WorkspaceUsers.get(workspace_id, member_id)
+    form = NewMemberForm(http_request.form)
```
Comment (patricksmithdds) [inline atst/routes/workspaces.py]: Should this be `form = EditMemberForm()`?

File: atst/routes/workspaces.py [hunk 3]
```diff
@@ -119,3 +122,44 @@ def create_member(workspace_id):
...
+    return render_template(
+        "member_edit.html", form=form, workspace=workspace, member=member
+    )
+
+
+@bp.route(
+    "/workspaces/<workspace_id>/members/<member_id>/member_edit", methods=["POST"]
+)
+def update_member(workspace_id, member_id):
```
Comment (patricksmithdds) [inline atst/routes/workspaces.py]: We should enforce the Authorization check here as well.

File: templates/member_edit.html [hunk 4]
```diff
@@ -2,223 +2,141 @@
...
-        <option value="Billing Auditor" selected="selected">Billing Auditor</option>
-        <option value="CCPO">CCPO</option>
-        <option value="Developer">Developer</option>
-        <option value="Owner">Owner</option>
-        <option value="Security Auditor">Security Auditor</option>
-      </select>
-    </div>
+  <div class='panel member-card'>
+    <div class='member-card__header'>
+      <h1 class='member-card__heading'>Danny Knight</h1>
```
Comment (patricksmithdds) [inline templates/member_edit.html]: Can we put in the real member data now that we're actually passing the data in?

File: templates/member_edit.html [hunk 5]
```diff
@@ -2,223 +2,141 @@
...
 
+        <div class="usa-input member-card__input">
+          <label for="filter-status">Workspace Role</label>
+          {{ Selector(form.workspace_role) }}
```
Comment (patricksmithdds) [inline templates/member_edit.html]: I think some of the styles might need to be tweaked here. Previously this looked like:

image

But now I'm seeing this:

image

Also, when the selector is expanded, the role names should be bolded (see the selector on the new member form).

File: tests/domain/test_workspaces.py [hunk 6]
```diff
@@ -119,3 +119,38 @@ def test_need_permission_to_create_workspace_user():
...
+        "last_name": "User",
+        "email": "new.user@mail.com",
+        "workspace_role": "developer",
+        "dod_id": "1234567890",
+    }
+    member = Workspaces.create_member(owner, workspace, user_data)
+    role_name = "developer"
+
+    updated_member = Workspaces.update_member(owner, workspace, member, role_name)
+    assert updated_member.workspace == workspace
```
Comment (patricksmithdds) [inline tests/domain/test_workspaces.py]: The `role_name` isn't different from the original role, so I think we should select a different role to update it to and also add an assertion to verify that the member's role was updated.

File: templates/member_edit.html (hunk 5 above)
Comment (patricksmithdds) [inline templates/member_edit.html]: This should also pre-select the user's current role. Currently, it shows "select an option" regardless of what the user's role is.

File: atst/routes/workspaces.py (hunk 2 above)
Comment (montana-mil) [reply atst/routes/workspaces.py]: yes

File: atst/domain/workspaces.py [hunk 7]
```diff
@@ -81,6 +81,15 @@ def create_member(cls, user, workspace, data):
         )
         return workspace_user
 
+    @classmethod
+    def update_member(cls, user, workspace, member, role_name):
+        if not Authorization.has_workspace_permission(
+            user, workspace, Permissions.ASSIGN_AND_UNASSIGN_ATAT_ROLE
+        ):
+            raise UnauthorizedError(user, "update workspace member")
```
[the remark this answers is not in the document]
Comment (montana-mil) [reply atst/domain/workspaces.py]: oh good catch. @richard-dds we can didn't have any reason against using `check_workspace_permission`, right? I'm going to switch it to use that.

File: atst/models/workspace.py (hunk 1 above)
Comment (montana-mil) [reply atst/models/workspace.py]: yep. done!

File: atst/domain/workspaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (montana-mil) [reply atst/domain/workspaces.py]: actually there are a number of `has_workspace_permissions` calls in that file. Should I switch them all out? @dandds @richard-dds

File: templates/member_edit.html [hunk 8]
```diff
@@ -2,223 +2,140 @@
...
-  <div class="panel__heading panel__heading--tight">
-    <h2 class="h3">Manage Access <div class="subtitle">Grant access to an environment</div></h2>
+  <div class="panel">
+    <div class="panel__heading panel__heading--tight">
+      <h2 class="h3">Manage Access <div class="subtitle">Grant access to an environment</div></h2>
+    </div>
   </div>
-</div>
 
-<form class='search-bar'>
```
Comment (patricksmithdds) [inline templates/member_edit.html]: Looks like removing this form changed the way the search part renders:

image

Was there an issue with having multiple forms on the page?

Comment (montana-mil) [reply templates/member_edit.html]: that part isn't wired up at all. is it important to make it look nice?

**Pull request merged**